### PR TITLE
Abstract out some of the descriptor Span-parsing helpers

### PR DESCRIPTION
**Pull request by sipa**
As suggested here: https://github.com/bitcoin/bitcoin/pull/16800#issuecomment-531605482.

This moves the Span parsing functions out of the descriptor module, making them more easily usable for other parsers (in particular, in preparation for miniscript parsing).

### CODE REVIEW

**Comment by promag** [comment]
ACK c072446b5c7fbaabcd17494d1b593b1cb5ebc314, verified it's mostly moved code - new code is just forward declarations, license and include guards.

File: src/util/spanparsing.h [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_UTIL_SPANPARSING_H
+#define BITCOIN_UTIL_SPANPARSING_H
+
+#include <string>
+#include <span.h>
+
+/** Parse a constant. If successful, sp is updated to skip the constant and return true. */
+bool Const(const std::string& str, Span<const char>& sp);
```
Comment (laanwj) [inline src/util/spanparsing.h]: with C++ throwing everything into the global namespace by default, I think these functions names  (`Const`, `Func`, `Expr`) are too general

**Comment by laanwj** [comment]
Concept ACK.

Now that these functions are public, would be nice to have unit tests.

File: src/util/spanparsing.h (hunk 1 above)
Comment (MarcoFalke) [reply src/util/spanparsing.h]: Make it a `namespace span` :) ?

Comment (sipa) [reply src/util/spanparsing.h]: I've moved them into a namespace spanparsing, and added some documentation.

**Comment by theStack** [comment]
utACK 230d43f 5e69aee

> Now that these functions are public, would be nice to have unit tests.

I would be interested to work on that.

**Comment by sipa** [comment]
@theStack Be my guest!

**Comment by theStack** [comment]
> @theStack Be my guest!

Unit tests for the Span-parsing helpers are ready, will open a PR for them as soon as this one is merged in (I guess it's not possible to open a PR which depends on another PR, right?)

**Comment by sipa** [comment]
@theStack You can post a commit, and I can include it in my PR. You can also open a PR against my branch.

**Comment by fanquake** [comment]
> You can post a commit, and I can include it in my PR. 

👍 Then we can have the tests alongside the changes in the same PR. @theStack obviously your changes remain attributed to you.

**Comment by theStack** [comment]
@sipa @fanquake Okay, here we go: https://github.com/theStack/bitcoin/commit/f4db52f932bb18346fda2d3e1bb95290acacc12e

**Comment by laanwj** [comment]
ACK 5e69aeec3f2a0fafd5e591b7222716f00145761d
ACK f4db52f932bb18346fda2d3e1bb95290acacc12e

@theStack thanks !

**Comment by laanwj** [comment]
@sipa if you agree with @theStack's tests, please pull the commit in here

**Comment by sipa** [comment]
Will do next week.

**Comment by sipa** [comment]
Thanks, @theStack; I've included your commit.

**Comment by practicalswift** [comment]
The parsing heavy functions `Const`, `Func`, `Expr` and `Split` all seem like good candidates for fuzz testing (in addition to the ordinary unit testing)?

**Comment by MarcoFalke** [comment]
ACK bb36372b8f2bd675313ae8553ceb61f28c2c1afd

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK bb36372b8f2bd675313ae8553ceb61f28c2c1afd
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjmOwwAi3Ad7AG9map87wVhJiT68gHzVlnpfujDXkepHRMoqt3E2T9zQuFk1QvW
m11WLTcf7VpTrn+zgGJ90w9HW3ovvNivHCCqwds5R+Rh2MW3u2cQJ7DkRQH7VePJ
B+VLLRXXuCq98kzF7XGqViyqmIA10tA+0rZ2DcIH8jguRsg9mLPybDlN1z+8j2c/
CHvb99LxnKiCHd0z7NgWXr1JsPsh97W5pnDABMybEfG3Oblh0DD/JdVf/w6cdWQD
vD/zuy54Re0CEA8Y4CUviw0OtlfHdiRsqepumXACMxJTvA/EzxOIHpaQZpj/pDOL
YzxHjBmrV/I4fYekscpA+uaPVwKvzHnOEQtZIBFs2/jK1FoaXfBdzSYvorj9V4q5
+DwFPa4aye0lVslBo9sED6ehUzLsP50tF80XrLudTAEKTWjUAaIUf0Mfe/6y4Op3
+XeEvhjOr33N7f6w3a+jdX2DR0MsRMCLZ4FPlLatGVGCfqgPtIWcxi5Bxitmdrla
Gagqrt9v
=NYw8
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `fb19a80d964f70a8c4b0e02b5fb1088d6d3c529398648e56d30d9e2f6fb4ba33  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108fb19a80d964f70a8c4b0e02b5fb1088d6d3c529398648e56d30d9e2f6fb4ba33f010c7151fa2e875102be6c0dc64a7334d6808fff01032061841c66ae6216522fc14bd69074508f1045d9f5c87f00865caa1a13185c3d00083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff010780a6a7467931ee1226e83e7df85dc1908f1205674d9fbc7994cc6e7354bcc5c0b9a1a660feca11af34834ed971217ae6931da08f1045d9f5c87f0083e814551a880dedf0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010a21e24d6de49fec1269746a43be5870508f0206602f0bbcd9ec1c8b88cbadf772acb6199ba99706fc5b55ba76810a2ede9996f08f1045d9f5c88f008d5a7b898b654dcc50083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010e31b606255c3f73d4b2f457a0778ba3e08f0207c344e4ed05d6c853d803237d7e0d02f4c86063f651a0c82466331c0d5451e4408f1045d9f5c88f008fcaa3aa662c2c54f0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6d)

</details>

**Comment by MarcoFalke** [comment]
Pretty straightforward move-only commit and then a commit adding tests, going to merge now.

**Comment by promag** [comment]
reACK 230d43fdbc41b356700b0d8a6984d69e00279ade, utACK bb36372b8f2bd675313ae8553ceb61f28c2c1afd.

**Pull request merged**